pkg/circuitbreaker/circuit_breaker.go

```diff
@@ -0,0 +1,108 @@
+package circuitbreaker
+
+import (
+	. "github.com/idealo/aws-signing-proxy/pkg/logging"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/sony/gobreaker"
+	"go.uber.org/zap"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type CircuitBreaker struct {
+	breaker *gobreaker.CircuitBreaker
+}
+
+func NewCircuitBreaker(name string) *CircuitBreaker {
+
+	timeout := getTimeout()
+	failureThreshold := getFailureThreshold()
+
+	st := gobreaker.Settings{
+		Name:    name,
+		Timeout: timeout,
+		ReadyToTrip: func(counts gobreaker.Counts) bool {
+			return counts.ConsecutiveFailures > failureThreshold
+		},
+	}
+	return &CircuitBreaker{breaker: gobreaker.NewCircuitBreaker(st)}
+}
+
+func getFailureThreshold() uint32 {
+	var failures uint32 = 5
+	failuresStr, ok := os.LookupEnv("ASP_CIRCUIT_BREAKER_FAILURE_THRESHOLD")
+	if ok {
+		i, err := strconv.Atoi(failuresStr)
+		if err != nil {
+			Logger.Error("Failed parsing the circuit breaker failure count", zap.Error(err))
+			return failures
+		}
+		return uint32(i)
+	}
+	return failures
+}
+
+func getTimeout() time.Duration {
+	var timeout time.Duration
+	var err error
+
+	timeoutStr, ok := os.LookupEnv("ASP_CIRCUIT_BREAKER_TIMEOUT")
+	if ok {
+		timeout, err = time.ParseDuration(timeoutStr)
+		if err != nil {
+			Logger.Error("Failed parsing the circuit breaker timeout", zap.Error(err))
+			timeout = 0 // defaults to 60s
+		}
+	}
+	return timeout
+}
+
+var (
+	cbStateGauge   = promauto.NewGaugeVec(prometheus.GaugeOpts{Name: "auth_circuit_breaker_state", Help: "State of the authorization circuit breaker"}, []string{"state", "name"})
+	cbCounterGauge = promauto.NewGaugeVec(prometheus.GaugeOpts{Name: "auth_circuit_breaker_count", Help: "Circuit breaker request count"}, []string{"type"})
+)
+
+func (cb *CircuitBreaker) Execute(req func() (interface{}, error)) (interface{}, error) {
+
+	r, err := cb.breaker.Execute(req)
```

Comment:
I think with the r variable the request is meaning, but for me it is to shorten. Maybe we should rename it to the full varibale name request?